Create streamlined 'golden path' installer for new users (#1414)
LorenFrankLab/spyglass: docs/src/GettingStarted/DATABASE.md

1. **Install Docker Desktop:**
    ([macOS](https://docs.docker.com/desktop/install/mac-install/),
    [Windows](https://docs.docker.com/desktop/install/windows-install/), or
    [Linux](https://docs.docker.com/desktop/install/linux-install/))

2. **Start Docker Desktop** and ensure it's running

3. **Verify Compose is available:**

    ```bash
    docker compose --version
    # Should show: Docker Compose version ##.##.##
    ```

#### Setup

**Using installer (recommended):**

```bash
python scripts/install.py --docker  # Will auto-detect and use compose command
```

**Using Docker Compose directly:**

```bash
# From spyglass repository root
docker compose up -d
```

The default configuration uses:

- Port: 3306
- Password: tutorial
- Container name: spyglass-db
- Persistent storage: spyglass-db-data volume

#### Customization (Optional)

Create a `.env` file to customize settings:

```bash
# Copy and edit example
cp example.env .env
nano .env # or use your favorite editor
```

Available options:

```bash
# Change port if 3306 is in use
MYSQL_PORT=3307

# Change root password (for production)
MYSQL_ROOT_PASSWORD=your-secure-password

# Use different MySQL version

> MYSQL_IMAGE=datajoint/mysql:8.4

Reviewer: [Here](https://hub.docker.com/r/datajoint/mysql/tags) are the tags that would work

```suggestion
MYSQL_IMAGE=datajoint/mysql:8.0.18
```

  "loglevel": "INFO",
  "safemode": true,
  "fetch_format": "array"
}
```

**Key settings:**

- `database.host`: MySQL server hostname or IP
- `database.port`: MySQL port (default: 3306)
- `database.user`: MySQL username
- `database.password`: MySQL password
- `database.use_tls`: Use TLS/SSL encryption (recommended for remote)

### TLS/SSL Configuration

**When to use TLS:**

- ✅ Remote database connections
- ✅ Production environments
- ✅ When connecting over untrusted networks
- ❌ localhost connections
- ❌ Docker containers on same machine

**Enable TLS:**

```json
{
  "database.use_tls": true
}
```

**Custom certificate:**

```json
{
  "database.use_tls": {
    "ssl": {
      "ca": "/path/to/ca-cert.pem",
      "cert": "/path/to/client-cert.pem",
      "key": "/path/to/client-key.pem"
    }
  }
}
```

## Security Best Practices

### Development

For local development, simple credentials are acceptable:

- User: `root` or dedicated user
- Password: Simple but unique
- TLS: Disabled for localhost

### Production

For shared/production databases:

1. **Strong passwords:**

    ```bash
    # Generate secure password
    openssl rand -base64 32
    ```

> 2. **User permissions:**

Reviewer: ```suggestion
2. **User permissions:** Manage by database prefix
```

    ```sql
    -- Create user with specific database access
    CREATE USER 'spyglass'@'%' IDENTIFIED BY 'strong_password';

>     GRANT ALL PRIVILEGES ON spyglass_*.* TO 'spyglass'@'%';

Reviewer: Should this have a line for `common_`, `lfp_`, etc? Or some kind of note indicating that each would need to be granted in a prefix-managed usecase?

My `database_settings.py` file was not clear to one outside lab using Spyglass. It may be worth scrapping that in favor of `sql` files in some subdirectory

```suggestion
    GRANT ALL PRIVILEGES ON someprefix_*.* TO 'spyglass'@'%';
```

<details><summary>SQL script worth adding somewhere</summary>

```sql
-- -- CREATE ROLES ---
-- NOTE: After any operation involving roles/users, run: FLUSH PRIVILEGES;

-- Create Admin Role
CREATE ROLE IF NOT EXISTS 'dj_admin';
GRANT ALL PRIVILEGES ON `%`.* TO 'dj_admin';

-- Create User Role
CREATE ROLE IF NOT EXISTS 'dj_user';
GRANT SELECT ON `%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `common\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `spikesorting\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `decoding\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `position\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `linearization\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `ripple\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `lfp\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `waveform\_%`.* TO 'dj_user';
GRANT ALL PRIVILEGES ON `mua\_%`.* TO 'dj_user';
-- It's easier to revoke CREATE than grant all the other privileges
-- Revoke CREATE privileges disables non-admin users from creating new tables
REVOKE CREATE ON `common\_%`.* FROM 'dj_user';
REVOKE CREATE ON `decoding\_%`.* FROM 'dj_user';
REVOKE CREATE ON `lfp\_%`.* FROM 'dj_user';
REVOKE CREATE ON `linearization\_%`.* FROM 'dj_user';
REVOKE CREATE ON `mua\_%`.* FROM 'dj_user';
REVOKE CREATE ON `position\_%`.* FROM 'dj_user';
REVOKE CREATE ON `ripple\_%`.* FROM 'dj_user';
REVOKE CREATE ON `spikesorting\_%`.* FROM 'dj_user';
REVOKE CREATE ON `waveform\_%`.* FROM 'dj_user';

-- Create Guest Role
CREATE ROLE IF NOT EXISTS 'dj_guest';
GRANT SELECT ON `%`.* TO 'dj_guest';

-- Create Collaborator Role
CREATE ROLE IF NOT EXISTS 'dj_collab';
GRANT SELECT ON `%`.* TO 'dj_collab';

-- -- CREATE USERS ---
-- Create Admin user
CREATE USER IF NOT EXISTS 'ADMIN_USER'@'%' IDENTIFIED BY 'ADMIN_PASSWORD';
GRANT dj_admin TO 'ADMIN_USER'@'%';
FLUSH PRIVILEGES;

-- Create Normal user
CREATE USER IF NOT EXISTS 'NORMAL_USER'@'%' IDENTIFIED BY 'NORMAL_PASSWORD';
GRANT dj_user TO 'NORMAL_USER'@'%';
GRANT ALL PRIVILEGES ON `USER\_%`.* TO 'NORMAL_USER'@'%';
FLUSH PRIVILEGES;

-- Create Guest user
CREATE USER IF NOT EXISTS 'GUEST_USER'@'%' IDENTIFIED BY 'GUEST_PASSWORD';
GRANT USAGE ON *.* TO 'GUEST_USER'@'%';
GRANT SELECT ON `%`.* TO 'GUEST_USER'@'%';
GRANT 'dj_guest'@'%' to 'GUEST_USER'@'%';
FLUSH PRIVILEGES;

-- Create Collaborator user
CREATE USER IF NOT EXISTS 'COLLAB_USER'@'%' IDENTIFIED BY 'COLLAB_PASSWORD';
GRANT ALL PRIVILEGES ON `USER\\_%`.* TO 'COLLAB_USER'@'%';
GRANT SELECT ON `%`.* TO 'COLLAB_USER'@'%';
FLUSH PRIVILEGES;

-- --- RESETTING PASSWORDS FOR EXISTING USERS ---
use mysql;
ALTER USER this_user IDENTIFIED BY 'NEW_PASSWORD';
FLUSH PRIVILEGES;

-- --- LOCKING/UNLOCKING EXISTING USERS ---
ALTER USER 'user_name'@'host' ACCOUNT LOCK;
ALTER USER 'user_name'@'host' ACCOUNT UNLOCK;

-- -- DISABLING TABLE OPERATIONS VIA TRIGGERS ---
-- Create a trigger to prevent deletion of rows in some table
use some_schema;
CREATE TRIGGER prevent_delete_some_table
BEFORE DELETE ON `some_schema`.`some_table`
FOR EACH ROW
SIGNAL SQLSTATE '45000'
SET MESSAGE_TEXT = 'DELETE temporarily disabled for `some_schema`.`some_table`';
-- Turn off the trigger
DROP TRIGGER IF EXISTS prevent_delete_some_table;
```

</details>

    FLUSH PRIVILEGES;
    ```

3. **Enable TLS:**

    ```json
    {
      "database.use_tls": true
    }
    ```

4. **Network security:**

    - Use firewall rules
    - Consider VPN for remote access
    - Use SSH tunnels when appropriate

5. **Credential management:**

    - Never commit config files to git
    - Use environment variables for CI/CD
    - Consider secrets management tools

### File Permissions

> Protect configuration file:

Reviewer: ```suggestion
Protect configuration file by only allowing read/write access by you:
```

```bash
chmod 600 ~/.datajoint_config.json
```

## Multi-User Setup

For lab environments with shared database:

### Server-Side Setup

```sql
-- Create database prefix for lab
CREATE DATABASE spyglass_common;

-- Create users
CREATE USER 'alice'@'%' IDENTIFIED BY 'alice_password';
CREATE USER 'bob'@'%' IDENTIFIED BY 'bob_password';

-- Grant permissions
GRANT ALL PRIVILEGES ON spyglass_*.* TO 'alice'@'%';

> GRANT ALL PRIVILEGES ON spyglass_*.* TO 'bob'@'%';

Reviewer: With `spyglass` as the example prefix, I'd worry that this would imply this is 
a spyglass-managed schema. Instead, a more neutral example might be better.

```suggestion
CREATE DATABASE yourlab_common;

-- Create users
CREATE USER 'alice'@'%' IDENTIFIED BY 'alice_password';
CREATE USER 'bob'@'%' IDENTIFIED BY 'bob_password';

-- Grant permissions
GRANT ALL PRIVILEGES ON yourlab_*.* TO 'alice'@'%';
GRANT ALL PRIVILEGES ON yourlab_*.* TO 'bob'@'%';
```

FLUSH PRIVILEGES;
```

### Client-Side Setup

Each user creates their own config:

**Alice's config:**

```json
{
  "database.host": "lab-db.university.edu",
  "database.user": "alice",
  "database.password": "alice_password",
  "database.use_tls": true
}
```

**Bob's config:**

```json
{
  "database.host": "lab-db.university.edu",
  "database.user": "bob",
  "database.password": "bob_password",
  "database.use_tls": true
}
```

## Troubleshooting

### Cannot Connect

**Check MySQL is running:**

```bash
# Docker
docker ps | grep spyglass-db

# System service (Linux)
systemctl status mysql

# Homebrew (macOS)
brew services list | grep mysql
```

**Test connection:**

```bash
# With mysql client
mysql -h HOST -P PORT -u USER -p

# With Python
python -c "import datajoint as dj; dj.conn().ping()"
```

### Permission Denied

```sql

> -- Grant missing privileges

Reviewer: ```suggestion
```

GRANT ALL PRIVILEGES ON *.* TO 'user'@'host';
FLUSH PRIVILEGES;
```

### Port Already in Use

```bash
# Find what's using port 3306
lsof -i :3306
netstat -an | grep 3306

# Use different port
docker run -p 3307:3306 ...
# Update config with port 3307
```

### TLS Errors

```python
# Disable TLS for localhost
config = {"database.host": "localhost", "database.use_tls": False}
```

For more troubleshooting help, see [TROUBLESHOOTING.md](TROUBLESHOOTING.md).

## Advanced Topics

> ### Database Backup

Reviewer: We might delete sections in `docs/src/Management.md` that are redundant with this section
